uc-crux-llvm/README.md

```diff
@@ -0,0 +1,295 @@
+# UC-Crux-LLVM
+
+UC-Crux-LLVM is a tool for under-constrained symbolic execution of C programs.
+It can be used to find undefined behavior and failing assertions, or for simple
+functions, to formally verify the absence of such behaviors.
+
+**UC-Crux-LLVM is still in development.**
+
+## Demo
+
+<!-- NOTE(lb) These programs are in the test suite as double_free.c and not_double_free.c --->
+
+<!-- TODO(lb) The printouts from the tool leave a lot to be desired here... --->
+
+`uc-crux-llvm` can use symbolic execution to find the potential double-free in this program:
+```c
+#include <stdlib.h>
+void double_free(int* ptr, int x) {
+  if (x % 2 == 0) {
+    free(ptr);
+  }
+  if (x % 3 == 0) {
+    free(ptr);
+  }
+}
+```
+```
+$ uc-crux-llvm --entry-points double_free double_free.c
+[CLANG] clang "-c" "-DCRUCIBLE" "-emit-llvm" "-g" "-I" "test/programs" "-I" "/some/pathc/includes" "-O1" "-o" "crux-build/double_free.bc" "test/programs/double_free.c"
+[Crux] Attempting to prove verification conditions.
+[Crux] Attempting to prove verification conditions.
+[Crux] Results for double_free
+[Crux] Found likely bugs:
+Pointer freed twice
+```
+That's not too impressive, a simple linter might be able to find that bug. However, since `uc-crux-llvm` uses symbolic execution, it can precisely conclude that the following program _does not_ have a potential double-free (or _any_ other undefined behavior), provided that it's passed a non-null pointer:
+```c
+#include <stdlib.h>
+void not_double_free(int *ptr, int x) {
+  if (x % 2 == 0) {
+    printf("even!\n"); // needed to prevent Clang from optimizing away the if/else
```

Comment:
I'm curious about this... does clang realize these branches are exclusive and combines them together?  In that case, maybe we need a more compelling example.

Reply:
Well, `clang` is free to assume they're exclusive because they both free the pointer and it can assume that the program doesn't allow undefined behavior. I'm definitely open to using a different example - in particular, it might be nice to use something interprocedural.

Reply:
Yeah, let's see if we can come up with something that is a little more natural-looking, where we don't have to cleverly convince the compiler to leave in the bug we want to find.

Reply:
Done, the example now requires no massaging, and involves interprocedural reasoning.